RSDK-7757: Add SetRPM to controlled, encoded and basic gpio motor drivers

martha-johnston opened the pull request:
- add `SetRPM` to all gpio motor drivers (basic, encoded controlled)
- add deprecated comments to all gpio motor drivers
- remove extra `goForMath` function
- tested on rover motor

martha-johnston commented on components/motor/gpio/basic.go:
```diff
@@ -328,7 +329,25 @@ func (m *Motor) GoTo(ctx context.Context, rpm, positionRevolutions float64, extr
+		return err
+	}
+
+	powerPct := rpm / m.maxRPM
```
doesn't seem like this is a linear relationship, the calculated powerPct does not result in the input rpm. might try to look into a better way to do this but it's unencoded so there's not any good way to figure out what the motor is doing.

randhid replied:
it is unfortunately not. 
![image](https://github.com/viamrobotics/rdk/assets/35934754/5f8a529e-6a7c-4058-8290-7cd64bcef862)

martha-johnston replied:
mm yeah alright that's a bummer

randhid commented on components/motor/gpio/basic.go:
```diff
@@ -289,6 +289,7 @@ func (m *Motor) GoFor(ctx context.Context, rpm, revolutions float64, extra map[s
 		return errors.Wrap(err, "error in GoFor")
 	}
 
+	// Deprecated: setting revolutions == 0 will spin the motor indefinitely at the specified RPM
```
Actually log these as warning for the user.
``` suggestion
	m.logger.Warn(Deprecated: setting revolutions == 0 will spin the motor indefinitely at the specified RPM")
```

martha-johnston replied:
can I leave both the comment and add the warning log?

martha-johnston replied:
I figured I'd just add it at every mention of the deprecated behavior

randhid replied:
for sure. Note - for the less used drivers than gpio motor I personally think it's okay just to remove the behavior, especially after we change the wheeled base driver.

randhid replied:
Idea: we could implement the warning in `motor/client.go`, so we only have one place to remove it on 9/1. 

We should retain the behavior for implementing the gpio motor going forever with a request of revolutions = 0, but I'm not sure we should for the less-popular drivers. I don't think all of them implement this behavior anyway.

martha-johnston replied:
oh I like that. yeah they definitely all don't. that's one of the driver groupings I made so it should all be one ticket later on

martha-johnston replied:
> Idea: we could implement the warning in `motor/client.go`, so we only have one place to remove it on 9/1.

I think this will work for everything except when using the go sdk?

martha-johnston replied:
nvm it works with go sdk, doesn't work on app tho for some reason, I'll look into it more

martha-johnston replied:
follow up to this, still not sure why `motor/client.go` doesn't log warnings on app, so I added the individual warnings in the drivers

randhid commented on components/motor/gpio/controlled.go:
```diff
@@ -259,15 +259,43 @@ func (cm *controlledMotor) GoTo(ctx context.Context, rpm, targetPosition float64
 // negative the motor will spin in the forward direction.
-// If revolutions is 0, this will run the motor at rpm indefinitely
 // If revolutions != 0, this will block until the number of revolutions has been completed or another operation comes in.
+// Deprecated: If revolutions is 0, this will run the motor at rpm indefinitely.
```
add logger

randhid commented on components/motor/gpio/mathutils.go:
```diff
@@ -24,7 +24,7 @@ func sign(x float64) float64 { // A quick helper function
 }
 
-// If revolutions is 0, the returned wait duration will be 0 representing that
+// Deprecated: If revolutions is 0, the returned wait duration will be 0 representing that
```
these can stay ass comments.

randhid commented on components/motor/gpio/controlled.go:
```diff
@@ -259,15 +259,43 @@ func (cm *controlledMotor) GoTo(ctx context.Context, rpm, targetPosition float64
+
+	cm.loop.Resume()
+	// set control loop values
+	velVal := math.Abs(rpm * cm.ticksPerRotation / 60)
```
We're now calling `motor.ticksperRotation` in several places out of lock. This could introduce races with reconfiguration, or GoFor with SetRPM.

martha-johnston replied:
I'll go through and make sure it's being locked everywhere

martha-johnston commented:
went through the WIP branch you sent me @randhid and added some of the changes, but there's couple of things I want to call out.

first, I didn't remove the `position` function because I think it would just lead to a lot of replicated code and I don't see the problem with it, but I'm interested in hearing your reasoning to remove I might be missing why it's helpful

also, as I was going through to look at the lock/unlocks, I realized there's not a lot of consistency with when we're locking so I added a lot, but I think there's got to be a cleaner way to make sure everything is protected so I'm going to go through it again, but this is a starting point

martha-johnston commented on components/motor/gpio/encoded.go:
```diff
@@ -298,30 +281,32 @@ func (m *EncodedMotor) GoFor(ctx context.Context, rpm, revolutions float64, extr
+	defer m.mu.Unlock()
+	// cancel makeAdjustments if it already exists
+	if m.makeAdjustmentsDone != nil {
+		m.makeAdjustmentsDone()
```
are these something we usually care about locking for? it seems like we don't in other drivers, but I can't see why not?

randhid replied:
Depends on what's accessing the variable. Since this is populated in GoFor and is read from everywhere, I don't think there's a read/write race that we care about for offset, for example, which is suitable across API calls and multiple threads reading from it at the same time.  

This is probably a good addition, but we should work on mutex readability when we refactor. 

Mutexs aren't the only way to ensure thread safety too, you can look at the go examples for channels https://gobyexample.com/stateful-goroutines, or we could even store the simpler variables on the sturct in atomics. However, we'll keep the encoded motor's style because it's okay for this driver to rely on its mutexes; we'll just try to simplify the adjuster code further, and write more coverage tests for the math in the future.

randhid commented:
> went through the WIP branch you sent me @randhid and added some of the changes, but there's couple of things I want to call out.
> 
> first, I didn't remove the `position` function because I think it would just lead to a lot of replicated code and I don't see the problem with it, but I'm interested in hearing your reasoning to remove I might be missing why it's helpful
> 
> also, as I was going through to look at the lock/unlocks, I realized there's not a lot of consistency with when we're locking so I added a lot, but I think there's got to be a cleaner way to make sure everything is protected so I'm going to go through it again, but this is a starting point

There's a code simplification that I think may work. Essentially, every time we call position in our loop, we end up doing this calculation:
```go
deltaPos := lastTicks + m.offset - currTicks - m.offset
```

This canceling what position is trying to add in terms of offset from Encoder. Furthermore, we ultimately do the entire calculation in ticks, converting to/from revolutions with tickPerRotation unnecessarily. The motor's `Position` function is the only function that makes sound use of both `m.TicksPerRotation` and `m.offset` in the sense that it tells the user what the current revolutions are and what our offset is and presents that information. I don't think we need to care about adding the math in the actual control code. We might be able to remove the need for the `direction` parameter as well since we now can get the direction the `basic` motor is going from `IsPowered`, and without multiple math operations, we can directly compare the `goalPos` in Ticks with the currentPos from the encoder. 

All of this is to simplify the code further and avoid accessing variables at the start in multiple places when they are unnecessary. I'm not expecting you to do both the clean-up tickets and the SetRPM tickets in one go, but simpler and scope-constrained code is way easier to debug in terms of race conditions, which, I think, will help the inconsistent `IsPowered` ticket.

randhid commented on components/motor/gpio/encoded.go:
```diff
@@ -411,11 +419,13 @@ func (m *EncodedMotor) IsMoving(ctx context.Context) (bool, error) {
 func (m *EncodedMotor) Stop(ctx context.Context, extra map[string]interface{}) error {
-	if m.rpmMonitorDone != nil {
-		m.rpmMonitorDone()
+	m.mu.Lock()
```
I'm assuming you're not using 
```go m.mu.Lock()
defer m.mu.Unlock()
```
because something hung somewhere?

martha-johnston replied:
no this should be defer

martha-johnston replied:
changed it in the latest commit

martha-johnston commented:
okay everything is in a working state right now, I will do any future refactors in the next ticket after setRPM is done

martha-johnston commented:
actually going to remove unnecessary locks based on your latest comment just so it's cleaner, then I will stop refactoring in this ticket

randhid commented on components/motor/gpio/controlled.go:
```diff
@@ -30,8 +30,6 @@ func (cm *controlledMotor) SetState(ctx context.Context, state []*control.Signal
 // State gets the state of the motor for the built-in control loop.
 func (cm *controlledMotor) State(ctx context.Context) ([]float64, error) {
 	ticks, _, err := cm.enc.Position(ctx, encoder.PositionTypeTicks, nil)
-	cm.mu.RLock()
-	defer cm.mu.RUnlock()
```
to make your testing a bit easier, I think it's sufficient to only care about your mutex cleaning in encoded motor.

martha-johnston replied:
this hasn't affected my testing, I think everything works in this state for both encoded and controlled, but I'll double check again.

martha-johnston replied:
I'm using bucket rover where one motor has controls and the other doesn't so I can test both at the same time

randhid replied:
great make sure you do all the interrupting teest where:

- [ ] A GoFor interrupts a GoFor on the RC Card
- [ ] SetVelocity Interrupts MoveStraight on the RC Card
- [ ] motor.SetPower Interrupts GoFor on the RC Card
- [ ] base.SetPower Interrupts MoveStraight On the RC Card
- [ ] You can still shut down gracefully after each routine is cancelled, no threads remain after close.

martha-johnston replied:
second one doesn't work on stable so I'll try and fix that bug

randhid replied:
Add a new ticket and fix it then.

randhid commented on components/motor/gpio/encoded_test.go:
```diff
@@ -206,20 +206,20 @@ func TestEncodedMotor(t *testing.T) {
 
 		expectedGoalPos, expectedGoalRPM, expectedDirection = -4.0, -10.0, -1.0
-		goalPos, goalRPM, direction = m.goForMath(context.Background(), 10, -4)
+		goalPos, goalRPM, direction = encodedGoForMath(10, -4, 0, 1)
```
let's add test coverage for our math

- test the infinite case with positive and negative direction
- test two negative inputs 
- test negative rpm and positive speed
- test zero zero

martha-johnston commented:
testing one more time and then will merge
